File: src/org/genyris/io/ReaderInStream.java
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.io;

import java.io.IOException;
import java.io.PushbackReader;
import java.io.Reader;

import org.genyris.exception.GenyrisException;
import org.genyris.interp.Environment;

public class ReaderInStream implements InStream {

    private int _nextByte;
    private boolean _haveChar;
    private PushbackReader _reader;
    private int _lineCount;

    public ReaderInStream(Reader reader) {
        _reader = new PushbackReader(reader);
        _haveChar = false;
        _lineCount = 1;
    }

    public Reader getReader() {
        return _reader;
    }

    public void unGet(char x) throws LexException {
        char[] charArray = new char[1];
        charArray[0] = x;
        try {
            _reader.unread(charArray);
        } catch (IOException e) {
            throw new LexException(e.getMessage());
        }
    }

    public char readNext() {
        _haveChar = false;
        if (_nextByte == '\n') {
            _lineCount++;
        }
        return (char) _nextByte;
    }

    public boolean hasData() {
        try {
            if (_haveChar) {
                return true;
            }
            _nextByte = _reader.read();
        } catch (IOException e) {
            return false;
        }
        if (_nextByte == -1) {
            _haveChar = false;
            return false;
        } else {
            _haveChar = true;
            return true;
        }
    }

    public void close() throws GenyrisException {
        try {
            _reader.close();
        } catch (IOException e) {
            throw new GenyrisException(e.getMessage());
        }
    }

    public void resetAfterError() {
        _haveChar = false;
    }

    public void withinExpression(Environment env) {
    }

    public void beginningExpression() {
    }

    public int getLineNumber() {
        return _lineCount;
    }

}

File: src/org/genyris/io/ReaderInStream.kt
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.io

import org.genyris.exception.GenyrisException
import org.genyris.interp.Environment
import java.io.IOException
import java.io.PushbackReader
import java.io.Reader

class ReaderInStream(reader: Reader, filename: String?) : InStream {
    private var _nextByte = 0
    private var _haveChar: Boolean
    private val _reader: PushbackReader
    private var _lineCount: Int
    private val _filename: String?

    init {
        _reader = PushbackReader(reader)
        _haveChar = false
        _lineCount = 1
        _filename = filename
    }

    override fun getReader(): Reader {
        return _reader
    }

    @kotlin.Throws(LexException::class)
    override fun unGet(x: Char) {
        if (x == '\n') {
            _lineCount--
        }
        val charArray = CharArray(1)
        charArray[0] = x
        try {
            _reader.unread(charArray)
        } catch (e: IOException) {
            throw LexException(e.getMessage())
        }
    }

    override fun readNext(): Char {
        _haveChar = false
        if (_nextByte == '\n'.code) {
            _lineCount++
        }
        return _nextByte.toChar()
    }

    override fun hasData(): Boolean {
        try {
            if (_haveChar) {
                return true
            }
            if (!_reader.ready()) {
                return false
            }
            _nextByte = _reader.read()
        } catch (e: IOException) {
            return false
        }
        if (_nextByte == -1) {
            _haveChar = false
            return false
        } else {
            _haveChar = true
            return true
        }
    }

    @kotlin.Throws(GenyrisException::class)
    override fun close() {
        try {
            _reader.close()
        } catch (e: IOException) {
            throw GenyrisException(e.getMessage())
        }
    }

    override fun resetAfterError() {
        _haveChar = false
    }

    override fun withinExpression(env: Environment?) {
    }

    override fun beginningExpression() {
    }

    override fun getLineNumber(): Int {
        return _lineCount
    }

    override fun getFilename(): String? {
        return _filename
    }
}
